apm-agent-core/src/main/java/co/elastic/apm/agent/impl/ElasticApmTracer.java

```diff
@@ -317,6 +317,7 @@ public Span startSpan(AbstractSpan<?> parent, long epochMicros) {
         Transaction transaction = currentTransaction();
         if (transaction != null) {
             if (isTransactionSpanLimitReached(transaction)) {
+                logger.warn("Max spans ({}) for a transaction has been reached for transaction {}. Further spans will be ignored. See config param 'transaction_max_spans'.", coreConfiguration.getTransactionMaxSpans(), transaction);
```

Comment:
We can't log it at every span creation attempt. If the max is 500 and transactions have 1000 on average, we will get 500 log messages **for each transaction**. Maybe the simplest is log once (at most) per time frame, say- 5 minutes. WDYT?

Reply:
BTW, this method is invoked with a timestamp, so you can use it for that- will spare a couple of CPU cycles for not querying the system timing APIs 🤓

Reply:
Actually what might make more sense is to just move the check and log into `ElasticAPMTracer.endTransaction()` of the transaction. That way at least when the transaction is finished we can report out that the spans were not sent due to the max being met.

Reply:
We don't want to log for each transaction instance and even making sure we log for each transaction type feels overkill. All we want is make the user aware that there is a situation of losing data due to exceeding a maximum. Since such messages can get quickly lost over time as logs accumulate, I suggested logging every 5 minutes. Unfortunately, this implies we need to maintain a state. I don't like that too much as well, but I think it is fine in this case.

As long as transactions are shorter than 5 minutes, this suggested code have the same issue. The diff between the span timestamp and the transaction timestamp is irrelevant.

So I suggest you just introduce this new `lastSpanMaxWarningTimestamp` state and use it.

Two notes:
- Why modulus and not `if (epochMicros - lastSpanMaxWarningTimestamp > MAX_LOG_INTERVAL_MICRO_SECS`)?
- No need for `logger.isWarnEnabled()` - in this case it will cost more than it saves

Reply:
Thanks for clarifying. I'll implement it that way. 

Re: Why modulus.... no reason really. Should have probably been `>` you're right.

Reply:
Should be fixed based on fb95082